felix/fv/qos_controls_test.go

```diff
@@ -0,0 +1,186 @@
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"regexp"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/calico/felix/fv/infrastructure"
+	"github.com/projectcalico/calico/felix/fv/workload"
+	"github.com/projectcalico/calico/libcalico-go/lib/apiconfig"
+	api "github.com/projectcalico/calico/libcalico-go/lib/apis/v3"
+)
+
+var _ = infrastructure.DatastoreDescribe(
+	"_BPF-SAFE_ QoS controls tests",
+	[]apiconfig.DatastoreType{apiconfig.Kubernetes},
+	func(getInfra infrastructure.InfraFactory) {
+
+		var (
+			infra  infrastructure.DatastoreInfra
+			tc     infrastructure.TopologyContainers
+			topt   infrastructure.TopologyOptions
+			w      [2]*workload.Workload
+			cancel context.CancelFunc
+		)
+
+		BeforeEach(func() {
+			infra = getInfra()
+			topt = infrastructure.DefaultTopologyOptions()
+			tc, _ = infrastructure.StartNNodeTopology(1, topt, infra)
+
+			infra.AddDefaultAllow()
+
+			for ii := range w {
+				wIP := fmt.Sprintf("10.65.%d.2", ii)
+				wName := fmt.Sprintf("w%d", ii)
+				w[ii] = workload.Run(tc.Felixes[0], wName, "default", wIP, "8055", "tcp")
+				w[ii].ConfigureInInfra(infra)
+			}
+		})
+
+		AfterEach(func() {
+			tc.Stop()
+			infra.Stop()
+
+			if cancel != nil {
+				cancel()
+			}
+		})
+
+		Context("With bandwidth limits", func() {
+			var (
+				serverCmd *exec.Cmd
+			)
+
+			getQdisc := func() string {
+				out, err := tc.Felixes[0].ExecOutput("tc", "qdisc")
+				log.Infof("tc qdisc output:\n%v", out)
+				Expect(err).NotTo(HaveOccurred())
+				return out
+			}
+
+			It("should limit bandwidth correctly", func() {
+				By("Starting iperf3 server on workload 0")
+				serverCmd = w[0].ExecCommand("iperf3", "-s")
+				err := serverCmd.Start()
+				Expect(err).NotTo(HaveOccurred())
+
+				By("Running iperf3 client on workload 1")
+				out, err := w[1].ExecOutput("iperf3", "-c", w[0].IP, "-O5", "-J")
+				Expect(err).NotTo(HaveOccurred())
+				baselineRate, err := getRateFromJsonOutput(out)
+				Expect(err).NotTo(HaveOccurred())
+				log.Infof("iperf client rate with no limit (bps): %v", baselineRate)
+				// Expect the baseline rate to be much greater than the desired limit
+				Expect(baselineRate).To(BeNumerically(">=", 100000.0*100))
+
+				By("Setting 100kbps limit for ingress on workload 1")
+				w[1].WorkloadEndpoint.Spec.QoSControls = &api.QoSControls{
+					IngressBandwidth: 100000,
+					IngressBurst:     200000,
+				}
+				w[1].UpdateInInfra(infra)
+
+				By("Waiting for the config to appear in 'tc qdisc'")
+				// ingress config should be present
+				Eventually(getQdisc, "10s", "1s").Should(MatchRegexp(`qdisc tbf \d+: dev ` + regexp.QuoteMeta(w[1].InterfaceName) + ` root refcnt \d+ rate ` + regexp.QuoteMeta("100Kbit")))
+				// egress config should not be present
+				Eventually(getQdisc, "10s", "1s").ShouldNot(MatchRegexp(`qdisc tbf \d+: dev bwcali.* root refcnt \d+ rate ` + regexp.QuoteMeta("100Kbit")))
```

Comment:
No reason for Eventually here.  Should probably be just Expect.

Reply:
It does need `Eventually`, as it takes a quantum of time for the `tc qdisc` configuration to appear once the WorkloadEndpoint is updated. (granted, I didn't spend much time fine tuning this, 10s is probably too much)

Reply:
Oh right! I'll change the negative tests to Consistently. Sorry about that!

Reply:
Yes right, I'll change the appropriate ones to Consistently, sorry about that